src/irmin-pack/inode.ml

```diff
@@ -1103,52 +1103,86 @@ struct
     type proof = Node.proof [@@deriving irmin]
 
     module Proof = struct
-      let rec proof_of_concrete h : Concrete.t -> proof = function
-        | Blinded -> `Blinded (Lazy.force h)
-        | Values vs -> `Values (List.map Concrete.of_entry vs)
+      let rec proof_of_concrete :
+          type a. hash Lazy.t -> Concrete.t -> (proof -> a) -> a =
+       fun h proof k ->
+        match proof with
+        | Blinded -> k (`Blinded (Lazy.force h))
+        | Values vs -> k (`Values (List.map Concrete.of_entry vs))
         | Tree tr ->
             let proofs =
               List.fold_left
                 (fun acc (e : _ Concrete.pointer) ->
-                  let p = proof_of_concrete (lazy e.pointer) e.tree in
-                  let e = (e.index, p) in
-                  e :: acc)
+                  let hash = lazy e.pointer in
+                  proof_of_concrete hash e.tree (fun proof ->
+                      let e =
+                        match proof with
+                        | `Inode (_, [ (i, p) ]) -> (e.index :: i, p)
```

Comment:
Shouldn't a `List.rev` be called on the `in list` at some point?

Reply:
It's not necessary here as we are going out of the recursion here (and so the list is in the right order already). In any case, the new test is testing this and it seems to pass :-)